packages/core/content-releases/server/src/services/release.ts

```diff
@@ -161,6 +164,81 @@ const createReleaseService = ({ strapi }: { strapi: LoadedStrapi }) => ({
 
     return contentTypesMeta;
   },
+  async publish(releaseId: PublishRelease.Request['params']['id']) {
+    // We need to pass the type because entityService.findOne is not returning the correct type
+    const releaseWithEntries = (await strapi.entityService.findOne(RELEASE_MODEL_UID, releaseId, {
+      populate: {
+        actions: {
+          populate: {
+            entry: true,
+          },
+        },
+      },
+    })) as unknown as Release;
+
+    if (!releaseWithEntries) {
+      throw new errors.NotFoundError(`No release found for id ${releaseId}`);
+    }
+
+    if (releaseWithEntries.releasedAt) {
+      throw new errors.ValidationError('Release already published');
+    }
+
+    if (releaseWithEntries.actions.length === 0) {
+      throw new errors.ValidationError('No entries to publish');
+    }
+
+    /**
+     * We separate publish and unpublish actions group by content type
+     */
+    const actions: {
+      [key: UID.ContentType]: {
+        publish: ReleaseAction['entry'][];
+        unpublish: ReleaseAction['entry'][];
+      };
+    } = {};
+    for (const action of releaseWithEntries.actions) {
+      const contentTypeUid = action.contentType;
+
+      if (!actions[contentTypeUid]) {
+        actions[contentTypeUid] = {
+          publish: [],
+          unpublish: [],
+        };
+      }
+
+      if (action.type === 'publish') {
+        actions[contentTypeUid].publish.push(action.entry);
+      } else {
+        actions[contentTypeUid].unpublish.push(action.entry);
+      }
+    }
+
+    const entityManagerService = strapi.plugin('content-manager').service('entity-manager');
+
+    await strapi.db.transaction(async () => {
```

Comment:
I think we could add a small comment explaining the transaction. Something like:

```suggestion
		// Only publish the release if all action updates are applied successfully to their entry, otherwise leave everything as is
    await strapi.db.transaction(async () => {
```

Also I see it's experimental: https://contributor.strapi.io/docs/core/database/transactions